BDMS-112 Add missing fields to `Thing` model, schema, and tests

ksmuczynski opened the pull request:
<!--
- BDMS-112 Add missing fields to `Thing` model, schema, and tests
-->
### **Why**
_This PR addresses the following problem / context:_
- The `Thing` model needed to inherit the new tables that have been created.
- The `Thing` model needed a `first_visit_date` field to store the date of NMBGMR's first recorded interaction with a specific Thing.
- A `well_purpose` field was needed to capture the primary function of a well (Monitoring, Irrigation, etc)

### **How**
_Implementation summary - the following was changed / added / removed:_
- New `StatusHistoryMixin` and `PermissionMixin` were imported and inherited by the `Thing` model.
- Relationships were added and updated in the `Thing` model and related models.
- Syntax/notation was updated to reflect the SQLAlchemy 2.0 style.
- A `first_visit_date` field was added.
- A `well_purpose` field was added.
- A `well_type` field existed in schema files, test files, and the lexicon.json. The `well_type` field served the same purpose as the `well_purpose` field, despite not existing in the `Thing` model. Instances of `well_type` were replaced by `well_purpose` based on AMP feedback received on 2025-08-29.

### **Notes**
_Any special considerations, workarounds, or follow-up work to note?_
- Script files were not updated.

ksmuczynski committed:
feat: import mixins into `Thing` model

ksmuczynski committed:
reformat: updated fields to comform to SQLAlchemy 2.0 notation. Reorganized code into labeled sections

ksmuczynski committed:
reformat: rename code sections for clarity

ksmuczynski committed:
reformat: move well-related and spring-related columns under the `columns` code section

ksmuczynski committed:
feat: add versioning to Thing table

ksmuczynski committed:
refactor: add descriptions to relationships in `Thing` model

feat: add missing relationships to `Thing` model

ksmuczynski committed:
refactor: update relationships in `Thing` and `Group` tables.

feat: add relationships to `GroupThingAssociation` table.

feat: add proxy to `Group` table

ksmuczynski committed:
feat: add `thing` relationship to `WellScreen` model

ksmuczynski committed:
feat: add `screens` relationship to `Thing` model

ksmuczynski committed:
refactor: update `screens` relationship to `Thing` model

ksmuczynski committed:
refactor: add 'passive_deletes=True' parameter to necessary relationships in the Thing model

ksmuczynski committed:
refactor: add 'ondelete=CASCADE' to thing_id foreign key

ksmuczynski committed:
refactor: add 'passive_deletes=True' to `asset_associations` relationship in Thing model

ksmuczynski committed:
refactor: move `ondelete` parameter into foreign key constraint in Deployment model

ksmuczynski committed:
refactor: add TYPE_CHECKING to `Contact` model to mitigate circular references

ksmuczynski committed:
feat: add `first_visit_date` field to the `Thing` model

ksmuczynski committed:
refactor: rename `well_type` field to `well_purpose` as decided with AMP folks on 2025-08-29.

ksmuczynski committed:
feat: add `first_visit_date` field to applicable tests in `test.thing.py`

ksmuczynski committed:
feat: add `first_visit_date` field to `tests/conftest.py`

ksmuczynski committed:
refactor: correct polymorphic parent relationship on `Permission` table from `foreign(permissible_id) == Thing.thing_id` to `foreign(permissible_id) == Thing.id`

ksmuczynski committed:
fix(models): Correct primaryjoin condition in `PermissionMixin` and `StatusHistoryMixin`

The previous logic dynamically generated a primary key name based on the
class name (e.g., "thing_id"), which failed for models using the
AutoBaseMixin that provides a standard "id" column.

This change modifies the f-string to use the correct ".id" attribute,
resolving the AttributeError during mapper configuration.

ksmuczynski committed:
feat: add `permission` and `status` history models to `db/__init__.py`

ksmuczynski committed:
fix(models): Use string-based join in Permission model

The polymorphic relationships `_thing_target` and `_location_target`
caused a `NameError` at runtime because they referenced the `Thing` and
`Location` models directly. These models were only imported within a
`TYPE_CHECKING` block to avoid circular dependencies, making them
unavailable when the code was executed.

This commit refactors the `primaryjoin` condition to be a string. This
allows SQLAlchemy to defer the evaluation of the join until all models
are fully loaded, resolving the `NameError` and making the relationship
definitions more robust.

ksmuczynski committed:
fix(models): Correct ForeignKey reference for `contact_id` in `Permission` model

jacob-a-brown commented on api/search.py:
```diff
@@ -103,7 +103,7 @@ def make_well_response(thing: Thing) -> dict:
             thing,
             {
-                "well_type": thing.well_type,
+                "well_purpose": thing.well_type,
```
`well_type` was changed to `well_purpose` in the updates, so `thing.well_type` should change to `thing.well_purpose` here

jacob-a-brown commented on db/contact.py:
```diff
@@ -30,8 +35,12 @@ class ThingContactAssociation(Base, AutoBaseMixin):
+    )
+    thing: Mapped["Thing"] = relationship(
+        "Thing", back_populates="contact_associations"
+    )  # noqa: F821
```
`# noqa: F821` can be removed now that we are using `if TYPE_CHECKING`

jacob-a-brown commented on db/deployment.py:
```diff
@@ -4,6 +4,7 @@
 at which Thing and for what period of time.
 """
 
+from tkinter.constants import CASCADE
```
`tkinter` is a GUI library for Python. Is this an appropriate use of it? Or, if there are cascades that are commonly used they can be defined in `constants.py` in the root directory (and then imported where needed)

jirhiker replied:
this should be removed.   like the other instances of `ondelete`,  CASCADE should be a str e.g.  `ondelete="CASCADE"`

jacob-a-brown commented on db/thing.py:
```diff
@@ -30,104 +36,189 @@
 
-    well_construction_notes = Column(String(250))
+    # One To-Many: A Thing can be in many Groups over time.
+    thing_groups: Mapped[List["GroupThingAssociation"]] = relationship(
+        "GroupThingAssociation",
+        back_populates="thing",
+        cascade="all, delete-orphan",
+        passive_deletes=True,
+    )
```
this is called `thing_associations` in the `Group` table. For consistent style and naming should this be called `group_associations` here?

ksmuczynski replied:
Yes, good catch. `thing_association` is also more descriptive of the relationship. I will update this relationship name and its reference in the `groups` proxy.

jacob-a-brown commented on db/thing.py:
```diff
@@ -30,104 +36,189 @@
+        nullable=True,
+        comment="A controlled vocabulary field defining the type of infrastructure (e.g., 'Well', 'Spring', 'Stream Gauge').",
+    )
+    first_visit_date: Mapped[Date] = mapped_column(
+        Date,
+        nullable=True,
+        comment="The date of NMBGMR's first recorded interaction with this specific `Thing`.",
+    )
```
this could be an AMP question: they've been calling it `inventory date` for the longest time (i.e. `well inventory form`). Should this correspond with that vocabulary?

ksmuczynski replied:
They can keep referring to it as `inventory date` if they want, but in my opinion, `first_visit_date` is less ambiguous and more accurately describes what the field is storing. I think table and field names should be as explicit as possible.

jacob-a-brown commented on db/thing.py:
```diff
@@ -30,104 +36,189 @@
-    deployments: Mapped[List["Deployment"]] = relationship(
-        "Deployment", back_populates="thing", cascade="all, delete-orphan"
+    # Proxy to directly access the Contact objects associated with this Thing
+    contacts: AssociationProxy[list["Contact"]] = association_proxy(  # noqa: F821
```
`# noqa: F821` can be removed since `TYPE_CHECKING` is invoked.

jacob-a-brown commented on tests/conftest.py:
```diff
@@ -51,6 +51,7 @@ def water_well_thing(location):
+            first_visit_date="2023-03-03",
             thing_type="water well",
             release_status="draft",
             well_type="Production",
```
this needs to change to `well_purpose` for revisions

jacob-a-brown commented on db/thing.py:
```diff
@@ -30,104 +36,189 @@
+        nullable=True,
+        comment="A controlled vocabulary field defining the type of infrastructure (e.g., 'Well', 'Spring', 'Stream Gauge').",
+    )
+    first_visit_date: Mapped[Date] = mapped_column(
+        Date,
+        nullable=True,
+        comment="The date of NMBGMR's first recorded interaction with this specific `Thing`.",
+    )
```
should this be nullable?

ksmuczynski replied:
Yes. A NULL value here would indicate that the date of NMBGMR's first visit is unknown, which is a possibility, particularly with legacy data.

jacob-a-brown commented on schemas/thing.py:
```diff
@@ -46,14 +48,15 @@ class CreateBaseThing(BaseCreateModel):
     location_id: int | None = None  # Optional location ID for the thing
     group_id: int | None = None  # Optional group ID for the thing
     name: str  # Name of the thing
+    first_visit_date: date | None = None  # Date of NMBGMR's first visit
```
This should be `PastDate` to ensure that the date is not in the future. Also, should this be nullable when created?

jacob-a-brown replied:
https://docs.pydantic.dev/2.2/usage/types/datetime/

ksmuczynski replied:
`PastDate `makes sense, will update.

Yes, I think it should be nullable.  A NULL value would indicate that the date of NMBGMR's first visit is unknown, which is a possibility, particularly with legacy data.

jacob-a-brown replied:
We can allow it to be nullable in the data transfer, but should it be required for new records? If so, then it shouldn't be nullable in `CreateBaseThing` otherwise they'll be allowed to submit NULL values for new records

ksmuczynski replied:
Oh, I see. No, you are right, it should be required for new records. Updating this in the `schema/thing.py` file will only affect what the user is able to do, is that correct? The database will still be able to store NULL values for the legacy data?

jacob-a-brown replied:
That's correct, this will only effect incoming data. We don't use this for the transfer of legacy data. Since it is still nullable in the database it will be able to store the legacy NULL values

jacob-a-brown commented on schemas/thing.py:
```diff
@@ -93,6 +96,7 @@ class BaseThingResponse(BaseResponseModel):
     name: str
     thing_type: str
     active_location: LocationResponse | None = None
+    first_visit_date: date | None = None
```
should this field be nullable?

ksmuczynski replied:
Yes, I think so? If a user requests legacy data information from a Thing and its first visit date is unknown, shouldn't NULL be returned?

ksmuczynski committed:
feat: Add `Permission` relationship to `Contact` table

jacob-a-brown commented on schemas/thing.py:
should `first_date_visit` be changeable so people can update that field if they made a mistake during input? or found better documentation

ksmuczynski replied:
Yes, I think it should be editable. The following is what I have in the UPDATE schema - is this correct?
`first_visit_date: date | None = None `

jacob-a-brown replied:
Correct! Though it should be `PastDate` so the user can't enter a future date.

There's another style we are going to adopt for update that makes it so fields don't need to be in the payload but if they are present they must be of a certain type, but that's a future us problem.

ksmuczynski committed:
refactor: Add 'cascade="all, delete, delete-orphan"' to phone, emails, and addresses relationships in `Contact` model

ksmuczynski committed:
refactor: reorganize `Contact` model

Columns, relationships, and proxies were randomly added to the `Contact` model. It has been reorganized so like things are grouped together under a designed header.

ksmuczynski committed:
fix(models): Explicitly mark foreign keys in polymorphic joins

SQLAlchemy was raising an `ArgumentError` because it could not infer
the direction of the one-to-many polymorphic relationships defined in
the `StatusHistoryMixin` and `PermissionMixin`.

Without a formal `ForeignKey` constraint on the child columns
(`statusable_id` and `permissible_id`), SQLAlchemy could not determine
which side of the join was the foreign key.

This commit resolves the ambiguity by wrapping these child columns with
the `foreign()` annotation in their respective `primaryjoin` conditions.
This explicitly marks them as the foreign-key side of the join,
allowing the mapper to configure correctly.

jirhiker commented on db/deployment.py:
```diff
@@ -24,7 +25,9 @@ class Deployment(Base, AutoBaseMixin, ReleaseMixin):
     thing_id: Mapped[int] = mapped_column(
-        Integer, ForeignKey("thing.id"), nullable=False
+        Integer,
+        ForeignKey("thing.id", ondelete=CASCADE),
```
like the other instances of `ondelete=...` this should be `ondelete="CASCADE"`.  And then `from tkinter.constants import CASCADE` can be removed

ksmuczynski committed:
feat(models): Update column name in the `search_vector`.

The `Thing` model's search_vector referenced the
`well_casing_description` column, but it was not defined in the
schema, causing an `AttributeError` on startup.

The `well_casing_description` column was renamed `well_casing_material` so the search_vector was updated accordingly.

ksmuczynski committed:
fix(model): resolve errors to allow tests to run without issue.

`first_visit_date` was missing from the `schemas/thing.py` UPDATE model. It has been added.

Renamed instances of 'well_type' to 'well_purpose'

Commented out the `Thing.description` field

ksmuczynski committed:
refactor: update instance of `well_type` to `well_purpose`.

The `well_type` property was renamed to `well_purposed` in the `make_well_response` function.

ksmuczynski committed:
refactor: remove `tkinter` import

Instances of ondelete, CASCADE should be a str e.g. ondelete="CASCADE". The CASCADE from tkinter and tkinter itself was removed.

ksmuczynski committed:
refactor (schema): update data type

`first_visit_date` field should be of type `PastDate` to ensure a future date is not entered.

ksmuczynski committed:
refactor (schema): update `first_visit_date` field to be required (non-nullable)

When creating a new `Thing` record, users should be required to enter a value for `first_visit_date`. This applies to new records being created - the database itself will still allow NULLS in order to accommodate legacy data.

ksmuczynski committed:
refactor (model): update relationship name

The `thing_groups` relationship was updated to `group_associations` for clarity. It more descriptive of the relationship and is also consistent with the naming convention used in the group table (`group_associations)`.

The back_populates property reference in the group.thing relationship was updated from "thing_groups" to "group_associations" to reflect the update to the relationship name in the Thing table.

ksmuczynski committed:
refactor (model): remove # noqa: F821 calls

# noqa: F821 calls were removed since TYPE_CHECKING is invoked.

jacob-a-brown commented on db/contact.py:
```diff
@@ -41,53 +52,64 @@ class Contact(Base, AutoBaseMixin, ReleaseMixin):
-    # Proxy to directly access the FieldEvent objects in which this Contact participated.
-    # fmt: off
-    field_event_contact_associations: Mapped[list["FieldEventContactAssociation"]] = (  # noqa: F821
+    # One-To-Many: A Contact can participate in many Field Events.
+    field_event_contact_associations: Mapped[list["FieldEventContactAssociation"]] = (
```
After this gets merged I'll update my PR (171) to prevent merge conflicts (`field_event_contact_assocations` --> `field_event_participants`)

ksmuczynski replied:
Updated the following instances: `FieldEventContactAssociation`--> `FieldEventParticipants`
`field_event_contact_associations `--> `field_event_participants`

jacob-a-brown commented on db/contact.py:
```diff
@@ -41,53 +52,64 @@ class Contact(Base, AutoBaseMixin, ReleaseMixin):
+        "thing_associations", "thing"
+    )
+
+    # Proxy to directly access the FieldEvent objects in which this Contact participated.
+    field_events: AssociationProxy[list["FieldEvent"]] = association_proxy(
+        "field_event_contact_associations", "field_event"
+    )
```
After this gets merged I'll update my PR (171) to prevent merge conflicts (`field_event_contact_assocations` --> `field_event_participants`)

jacob-a-brown commented on schemas/thing.py:
```diff
@@ -192,11 +196,12 @@ class UpdateThing(BaseUpdateModel):
     """
 
     name: str | None = None  # Optional name for the thing
+    first_visit_date: date | None = None  # Date of NMBGMR's first visit
```
This should be `PastDate` to prevent the user from entering future dates

jirhiker replied:
@ksmuczynski  see above

ksmuczynski replied:
@jirhiker Done!

ksmuczynski committed:
refactor: update instances of `FieldEventContactAssociation` to  `FieldEventParticipants`. Update instances of `field_event_contact_associations` to `field_event_participants`

ksmuczynski committed:
fix: resolve foreign key reference

The foreign key on  `field_event_contact_id` field was incorrectly named "field_event_participant.id". It was updated to "field_event_participants.id"

ksmuczynski committed:
refactor: update data type

Change data type for `first_visit_date` from `date` to `PastDate`

jacob-a-brown reviewed (approved):
looks good to me
